File: .claude/hooks/ruff-check.mjs
#!/usr/bin/env node
// @hookstack post-write-ruff-check
import { execSync } from "node:child_process";
// @hookstack post-write-ruff-check
// Analyse et auto-corrige le fichier Python avec ruff après écriture (PostToolUse Write|Edit)
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";

function defaultExec(cmd) {
	return execSync(cmd, { encoding: "utf8", stdio: "pipe", timeout: 15_000 });
}

export function run(input, { exec = defaultExec } = {}) {
	const filePath = input.tool_input?.file_path ?? input.tool_input?.path ?? "";
	if (!filePath.endsWith(".py")) return null;

	try {
		exec(`uv run ruff check --fix "${filePath}"`);
		return null;
	} catch (err) {
		const output = err.stdout?.toString() ?? "";
		return output ? { message: `[ruff-check] ${output.trim()}\n` } : null;
	}
}

/* v8 ignore next 5 */
if (process.argv[1] === fileURLToPath(import.meta.url)) {
	const input = JSON.parse(readFileSync(0, "utf8"));
	const result = run(input);
	if (result?.message) process.stderr.write(result.message);
}

File: .claude/hooks/ruff-check.py
#!/usr/bin/env python3
# @hookstack post-write-ruff-check
"""Formats (silent) then lints and auto-fixes the Python file with ruff after
write (PostToolUse Write|Edit) — merge of ruff-format + ruff-check."""
import json
import subprocess
import sys


def _exec(cmd, timeout=15):
    return subprocess.run(
        cmd, shell=True, capture_output=True, text=True, timeout=timeout, check=True
    )


def run(input_data, exec_cmd=None):
    if exec_cmd is None:
        exec_cmd = _exec
    tool_input = input_data.get("tool_input") or {}
    file_path = tool_input.get("file_path") or tool_input.get("path") or ""
    if not file_path.endswith(".py"):
        return None

    # Format silencieux (non bloquant), puis lint --fix (erreurs remontées).
    try:
        exec_cmd(f'uv run ruff format "{file_path}"')
    except Exception:
        pass

    try:
        exec_cmd(f'uv run ruff check --fix "{file_path}"')
    except subprocess.CalledProcessError as err:
        output = err.stdout or ""
        return {"message": f"[ruff-check] {output.strip()}\n"} if output else None
    return None


if __name__ == "__main__":
    data = json.load(sys.stdin)
    result = run(data)
    if result and result.get("message"):
        sys.stderr.write(result["message"])
